File: lib/backtest.ts
// Episode runner shared by training and evaluation.

import type { QAgent } from "./agent";
import type { Trade } from "./broker";
import { type OptionsEnv, type State } from "./env";
import type { RsiStrategy } from "./strategy";

export interface EquityPoint {
  dayIndex: number;
  date: string;
  spot: number;
  equity: number;
}

export interface EpisodeResult {
  finalEquity: number;
  totalReward: number;
  trades: Trade[];
  equityCurve: EquityPoint[];
  nBuys: number;
  nSells: number;
  expiredItm: number;
  expiredOtm: number;
}

export type Policy = (state: State, env: OptionsEnv) => number;

export function agentPolicy(agent: QAgent, explore: boolean): Policy {
  return (state) => agent.act(state, explore);
}

export function rsiPolicy(strategy: RsiStrategy): Policy {
  return (state, env) => strategy.act(state, env);
}

export function runEpisode(
  env: OptionsEnv,
  policy: Policy,
  opts: { learn?: QAgent } = {},
): EpisodeResult {
  let state = env.reset();
  const firstBar = env.series.bars[env.warmup];
  const equityCurve: EquityPoint[] = [
    {
      dayIndex: firstBar?.t ?? 0,
      date: firstBar?.date ?? "",
      spot: firstBar?.close ?? 0,
      equity: env.equity,
    },
  ];
  let totalReward = 0;
  let done = false;

  while (!done) {
    const action = policy(state, env);
    const out = env.step(action);
    if (opts.learn) {
      opts.learn.update(state, action, out.reward, out.state, out.done);
    }
    state = out.state;
    totalReward += out.reward;
    equityCurve.push({
      dayIndex: out.info.dayIndex,
      date: out.info.date,
      spot: out.info.spot,
      equity: out.info.equity,
    });
    done = out.done;
  }

  const trades = env.broker.trades;
  return {
    finalEquity: env.equity,
    totalReward,
    trades,
    equityCurve,
    nBuys: trades.filter((t) => t.kind === "BUY").length,
    nSells: trades.filter((t) => t.kind === "SELL").length,
    expiredItm: trades.filter((t) => t.kind === "EXPIRE" && t.note === "ITM")
      .length,
    expiredOtm: trades.filter((t) => t.kind === "EXPIRE" && t.note === "OTM")
      .length,
  };
}

File: tradebot/backtest.py
"""Episode runner used by both training and evaluation."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, List

from .agent import QAgent
from .env import OptionsEnv
from .strategy import RSIStrategy


@dataclass
class EpisodeResult:
    final_equity: float
    total_reward: float
    n_trades: int
    n_buys: int
    n_sells: int
    n_expired_itm: int
    n_expired_otm: int
    equity_curve: List[float] = field(default_factory=list)


def run_episode(
    env: OptionsEnv,
    policy: Callable[[tuple, OptionsEnv], int],
    *,
    learn_agent: QAgent | None = None,
) -> EpisodeResult:
    """Run one episode. If ``learn_agent`` is provided, Q-values are updated."""
    state = env.reset()
    equity_curve: List[float] = [env.equity]
    total_reward = 0.0
    done = False

    while not done:
        action = policy(state, env)
        next_state, reward, done, info = env.step(action)
        if learn_agent is not None:
            learn_agent.update(state, action, reward, next_state, done)
        state = next_state
        total_reward += reward
        equity_curve.append(info["equity"])

    trades = env.broker.trades if env.broker else []
    return EpisodeResult(
        final_equity=env.equity,
        total_reward=total_reward,
        n_trades=len(trades),
        n_buys=sum(1 for t in trades if t.kind == "BUY"),
        n_sells=sum(1 for t in trades if t.kind == "SELL"),
        n_expired_itm=sum(1 for t in trades if t.kind == "EXPIRE" and t.note == "ITM"),
        n_expired_otm=sum(1 for t in trades if t.kind == "EXPIRE" and t.note == "OTM"),
        equity_curve=equity_curve,
    )


def agent_policy(agent: QAgent, *, explore: bool):
    def _p(state, _env):
        return agent.act(state, explore=explore)
    return _p


def rsi_policy(strategy: RSIStrategy):
    def _p(state, env):
        return strategy.act(state, env)
    return _p
